File: iOS/TreasureHunt/TreasureHuntObjc/LocationManager.m
//
//  LocationManager.m
//  TreasureHuntObjc
//
//  Created by Seamus on 05/11/2016.
//
//

#import "LocationManager.h"

@implementation LocationManager

- (id)init
{
    self = [super init];
    if (self) {
        self.locationManager = [[CLLocationManager alloc] init];
        [self.locationManager setDelegate:self];
        [self.locationManager setDesiredAccuracy:kCLLocationAccuracyThreeKilometers];
        [self.locationManager requestWhenInUseAuthorization];
    }
    return self;
}

-(void) requestLocation {
    [self.locationManager requestLocation];
}

-(void) locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    CLLocation* location = locations.lastObject;
    if (location != nil) {
        [_delegate didUpdateLocation:location];
    } else {
        [_delegate didFail];
    }
}

-(void) locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    [_delegate didFail];
}

@end

File: iOS/TreasureHunt/TreasureHunt/LocationManager.swift
//
//  LocationManager.swift
//  TreasureHunt
//
//  Created by Tim Colla on 05/11/2016.
//  Copyright © 2016 Marinosoftware. All rights reserved.
//

//
//  MCASLocationManager.swift
//  MCAS Demo
//
//  Created by Tim Colla on 10/10/2016.
//  Copyright © 2016 Marinosoftware. All rights reserved.
//

import Foundation
import CoreLocation

protocol LocationManagerDelegate {
    func didUpdate(location: CLLocation)
    func didFail()
}

class LocationManager : NSObject, CLLocationManagerDelegate {
    let locationManager = CLLocationManager()
    var delegate : LocationManagerDelegate?
    
    override init() {
        super.init()
        
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyThreeKilometers
        
        locationManager.requestWhenInUseAuthorization()
    }
    
    func requestLocation() {
        if #available(iOS 9.0, *) {
            locationManager.requestLocation()
        } else {
            // Fallback on earlier versions
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        if let location = locations.last {
            delegate?.didUpdate(location: location)
        } else {
            delegate?.didFail()
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        delegate?.didFail()
    }
}
